### PR TITLE
Make -s - read from stdin

### PR DESCRIPTION
**Pull request by b-r-o-c-k**
This is a continuation of #6299 with a fix I added for Windows.

### CODE REVIEW

**Review by justinmk — APPROVED** [review]

File: test/functional/core/main_spec.lua [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+      funcs.system(
+        {nvim_prog_abs(), '-u', 'NONE', '-i', 'NONE', '--headless',
+         '--cmd', 'set noswapfile shortmess+=IFW fileformats=unix',
+         '-s', dollar_fname, fname})
+      eq(0, eval('v:shell_error'))
+      local attrs = lfs.attributes(fname)
+      eq(#('100500\n'), attrs.size)
+    end)
+    it('does not crash after reading from stdin in non-headless mode', function()
+      if helpers.pending_win32(pending) then return end
```
Comment (justinmk) [inline test/functional/core/main_spec.lua]: does this still need to be skipped on windows?

File: src/nvim/os/fileio.c [hunk 2]
```diff
@@ -270,14 +282,17 @@ ptrdiff_t file_read(FileDescriptor *const fp, char *const ret_buf,
...
     const size_t rv_size = rbuffer_size(rv);
     if (rv_size > 0) {
       const size_t rsize = rbuffer_read(rv, buf, MIN(rv_size, read_remaining));
       buf += rsize;
       read_remaining -= rsize;
     }
-    if (fp->eof) {
+    if (fp->eof
+        // Allow only at most one os_read[v] call.
+        || (called_read && fp->non_blocking)) {
```
Comment (justinmk) [inline src/nvim/os/fileio.c]: @ZyX-I ok with this change?

File: test/functional/core/main_spec.lua (hunk 1 above)
Comment (b-r-o-c-k) [reply test/functional/core/main_spec.lua]: The test fails even though `:cq` works.
```
test\functional\ui\screen.lua:307: Row 2 did not match.
Expected:
  |^                                        |
  |*{1:~                                       }|
  |{1:~                                       }|
  |{1:~                                       }|
  |{1:~                                       }|
  |{2:[No Name]             0,0-1          All}|
  |                                        |
  |                                        |
Actual:
  |^                                        |
  |*                                        |
  |                                        |
  |                                        |
  |                                        |
  |                                        |
  |                                        |
  |                                        |
```

File: test/functional/core/main_spec.lua [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        {1:~                                       }|
+        {1:~                                       }|
+        {1:~                                       }|
+        {2:[No Name]             0,0-1          All}|
+                                                |
+                                                |
+      ]], {
+        [1] = {foreground = 4210943, special = Screen.colors.Grey0},
+        [2] = {special = Screen.colors.Grey0, bold = true, reverse = true}
+      })
```
Comment (janlazo) [inline test/functional/core/main_spec.lua]: Why is this necessary? This won't work on Windows until the TUI colors on cmd.exe or powershell .

File: src/nvim/os/fileio.c (hunk 2 above)
Comment (ZyX-I) [reply src/nvim/os/fileio.c]: OK.
